packages/react-keyring/src/Authenticator.tsx

```diff
@@ -0,0 +1,132 @@
+import React, {
+  useState, createContext, useContext, useCallback, useMemo
+} from 'react'
+import { useKeyring, KeyringContextState, KeyringContextActions } from './providers/Keyring'
+
+export type AuthenticatorContextState = KeyringContextState & {
+  /**
+   * email to be used to "log in"
+   */
+  email?: string
+  /**
+   * has the authentication form been submitted?
+   */
+  submitted: boolean
+  /**
+   * A callback that can be passed to an `onSubmit` handler to
+   * register a new space or log in using `email`
+   */
+  handleRegisterSubmit?: (e: React.FormEvent<HTMLFormElement>) => Promise<void>
+}
+
+export type AuthenticatorContextActions = KeyringContextActions & {
+  /**
+   * Set an email to be used to log in or register.
+   */
+  setEmail: React.Dispatch<React.SetStateAction<string>>
+}
+
+export type AuthenticatorContextValue = [
+  state: AuthenticatorContextState,
+  actions: AuthenticatorContextActions
+]
+
+export const AuthenticatorContext = createContext<AuthenticatorContextValue>([
+  {
+    spaces: [],
+    submitted: false
+  },
+  {
+    setEmail: () => { throw new Error('missing set email function') },
+    loadAgent: async () => { },
+    unloadAgent: async () => { },
+    resetAgent: async () => { },
+    createSpace: async () => { throw new Error('missing keyring context provider') },
+    setCurrentSpace: async () => { },
+    registerSpace: async () => { },
+    cancelRegisterSpace: () => { },
+    getProofs: async () => []
+  }
+])
+
+/**
+ * Top level component of the headless Authenticator.
+ *
+ * Must be used inside a KeyringProvider.
+ *
+ * Designed to be used by Authenticator.Form, Authenticator.EmailInput
+ * and others to make it easy to implement authentication UI.
+ */
+export function Authenticator (props: any): JSX.Element {
+  const [state, actions] = useKeyring()
+  const { createSpace, registerSpace } = actions
+  const [email, setEmail] = useState('')
+  const [submitted, setSubmitted] = useState(false)
+
+  const handleRegisterSubmit = useCallback(async (e: React.FormEvent<HTMLFormElement>) => {
```

Comment:
>  In both cases they add scope for subtle bugs since you have to manually specify dependencies.

Yep totally agree with this AND in this case because we're developing components designed to be used by outside developers I think it's more important to be careful with this stuff - this is a good example - if we don't use `useCallback` the `handleRegisterSubmit` function will be recreated on every render, triggering unnecessary re-renders in downstream components - in this case, because we pass this callback to the Context, it will potentially trigger unnecessary re-renders in any component that calls `useAuthenticator` (possibly even if they don't use `handleRegisterSubmit`, though I'm not 100% sure about that)

tldr: I agree with you philosophically but think we should use `useCallback` consistently for all callbacks in these libraries.

Reply:
I don't have a strong opinion about this. I'm happy to go with @travis' instinct here. If react users are going to complain at us for not doing it cos they all do, and it's not too onerous to apply it consistently then let's go! 

Alternatively, we could drop these optimisations, and wait for users (or better yet, our perf tests) to flag where we really need this. Optimising existing components can be a great "first contribution" task where folks want to get involved.